File: Classes/ThemeObject/UIImageView+LWTheme.m
//
//  UIImageView+LWTheme.m
//  LWTheme
//
//  Created by lalawue on 2019/12/13.
//  Copyright © 2019 lalawue. All rights reserved.
//

#import "UIImageView+LWTheme.h"
#import "LWTheme_Tools.h"

@interface LWImageColorItem : NSObject
@property (nonatomic, strong) UIImage *image;
@property (nonatomic, copy) NSString *tintColorKey;
@end

@implementation LWImageColorItem
@end

@implementation UIImageView (LWTheme)

- (NSString *)lw_imageName
{
    return (NSString *)[self lw_valueForSetter:@selector(setLw_imageName:)];
}

- (void)setLw_imageName:(NSString *)lw_imageName
{
    if (lw_imageName.length) {
        self.image = [self lw_imageNamed:lw_imageName];
        [self lw_storeValue:lw_imageName forSetter:@selector(setLw_imageName:)];
    }
}

- (void)lw_setImage:(UIImage *)image withTintColor:(NSString *)tintColor
{
    LWImageColorItem *item = [self lw_tintColorItem] ?: [LWImageColorItem new];
    item.image = image;
    item.tintColorKey = tintColor;
    if (tintColor.length) {
        UIColor *color = [self lw_colorForPattern:tintColor];
        self.image = [LWImageUtils fillImage:image withColor:color];
    } else {
        self.image = image;
    }
    [self lw_storeValue:item forSetter:@selector(lw_setTintColorItem:)];
}

#pragma mark - Apply

- (LWImageColorItem *)lw_tintColorItem
{
    return (LWImageColorItem *)[self lw_valueForSetter:@selector(lw_setTintColorItem:)];
}

- (void)lw_setTintColorItem:(LWImageColorItem *)item
{
    if (item.image && item.tintColorKey.length) {
        UIColor *color = [self lw_colorForPattern:item.tintColorKey];
        UIImage *image = [LWImageUtils fillImage:item.image withColor:color];
        [self setImage:image];
    }
}

@end

File: Classes/ThemeObject/UIImageView+LWTheme.swift
//
//  UIImageView+LWTheme.swift
//  LWThemeSwift
//
//  Created by lalawue on 2019/12/19.
//  Copyright © 2019 lalawue. All rights reserved.
//

import UIKit

fileprivate class ImageColorItem : NSObject {
    var image : UIImage? = nil
    var tintColorKey : String? = nil
}

extension UIImageView {
    
    @objc public var lw_imageName : String? {
        
        get {
            return lw_valueForSetter(#selector(setter: UIImageView.lw_imageName)) as? String
        }
        
        set (name) {
            self.image = lw_imageNamed(name)
            lw_storeValue(name, setter: #selector(setter: UIImageView.lw_imageName))
        }
    }
    
    @objc public func lw_setImage(_ image : UIImage?, tintColor : String?) {
        let item = lw_imageTintColorItem() ?? ImageColorItem()
        item.image = image
        item.tintColorKey = tintColor
        if (tintColor != nil) {
            let color = lw_colorForPattern(tintColor)
            self.image = image?.lw_tintImageWithColor(color, blendMode: CGBlendMode.destinationIn)
        } else {
            self.image = image
        }
        lw_storeValue(item, setter: #selector(UIImageView.lw_setImageTintColorItem(_:)))
    }
    
    //MARK: Apply
    
    fileprivate func lw_imageTintColorItem() -> ImageColorItem? {
        return lw_valueForSetter(#selector(UIImageView.lw_setImageTintColorItem(_:))) as? ImageColorItem
    }
    
    @objc fileprivate func lw_setImageTintColorItem(_ item : ImageColorItem) {
        if ((item.image != nil) && (item.tintColorKey != nil)) {
            let color = lw_colorForPattern(item.tintColorKey)
            self.image = item.image?.lw_tintImageWithColor(color, blendMode: CGBlendMode.destinationIn)
        }
    }
}
